File: binaryTreeZigzagLevelOrderTraversal.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > zigzagLevelOrder(TreeNode *root) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<vector<int> > result;
        if (root == NULL) {
            return result;
        }
        vector<int> v;
        vector<TreeNode*> stack;
        vector<TreeNode*> next_level_stack;
        stack.push_back(root);
        bool flag = false;
        while (!stack.empty()) {
            TreeNode* p = stack.front();
            stack.erase(stack.begin());
            if (flag) {
                v.insert(v.begin(), p->val);
            }
            else {
                v.push_back(p->val);
            }
            if (p->left != NULL) {
                next_level_stack.push_back(p->left);
            }
            if (p->right != NULL) {
                next_level_stack.push_back(p->right);
            }
            if (stack.empty()) {
                stack = next_level_stack;
                next_level_stack.clear();
                result.push_back(v);
                v.clear();
                flag = !flag;
            }
        }
        return result;
    }
};

File: binaryTreeZigzagLevelOrderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ArrayList<ArrayList<Integer>> zigzagLevelOrder(TreeNode root) {
        ArrayList<ArrayList<Integer>> result = new ArrayList<ArrayList<Integer>>();
        if (root == null) {
            return result;
        }
        ArrayList<Integer> v = new ArrayList<Integer>();
        ArrayList<TreeNode> stack = new ArrayList<TreeNode>();
        ArrayList<TreeNode> next_level_stack = new ArrayList<TreeNode>();
        stack.add(root);
        boolean flag = false;
        while (stack.size() != 0) {
            TreeNode p = stack.get(0);
            stack.remove(0);
            if (flag) {
                v.add(0, p.val);
            }
            else {
                v.add(p.val);
            }
            if (p.left != null) {
                next_level_stack.add(p.left);
            }
            if (p.right != null) {
                next_level_stack.add(p.right);
            }
            if (stack.size() == 0) {
                stack = next_level_stack;
                next_level_stack = new ArrayList<TreeNode>();
                result.add(v);
                v = new ArrayList<Integer>();
                flag = !flag;
            }
        }
        return result;
    }
}
